ratis-logservice/src/main/java/org/apache/ratis/logservice/server/LogStateMachine.java

```diff
@@ -263,26 +275,86 @@ public StateMachineStorage getStateMachineStorage() {
 
     ReadLogRequestProto msgProto = proto.getReadNextQuery();
     long startRecordId = msgProto.getStartRecordId();
-    int num = msgProto.getNumRecords();
+    int numRecordsToRead = msgProto.getNumRecords();
     Throwable t = verifyState(State.OPEN);
     List<byte[]> list = new ArrayList<byte[]>();
     LOG.info("Start Index: {}", startRecordId);
-    LOG.info("Total to read: {}", num);
+    LOG.info("Total to read: {}", numRecordsToRead);
+    long raftLogIndex = log.getStartIndex();
     if (t == null) {
-      for (long index = startRecordId; index < startRecordId + num; index++) {
+      // Seek to first entry
+      long logServiceIndex = 0;
```

Comment:
You're 100% correct. This is very inefficient right now, but I don't think we have anything better to do right now. I think to do something smarter, we would have to build some kind of indexing structure into our StateMachine which can safely jump to a certain point in the RAFT log and know the "recordId".

I'm not super-concerned with this right now as our primary use-case won't be doing random-reads over the log, instead sequential reads over most-all of the log.

Reply:
Josh, can we simplify this a little bit? Client gets Raft log ID one every append, for multi-get calls we can use this LogID (RAFT) as our start log ID to avoid iterating through all logs, then you can unwind log entries by using new logic, see above:
int numRecordsInAppend = append.getDataCount();

There is no need to iterate from the very beginning